docs/blog/posts/introducing-structured-outputs.md

```diff
@@ -0,0 +1,468 @@
+
+<!-- more -->
+
+### Limited Validation Feedback
+
+Validation is crucial for allowing models to correct their mistakes and improve their responses. Let's see a simple example where we want to extract a user's name in all uppercase.
+
+```python
+import openai
+from pydantic import BaseModel, field_validator
+
+
+class User(BaseModel):
+    name: str
+    age: int
+
+    @field_validator("name")
+    def ensure_uppercase(cls, v: str) -> str:
+        if not v.isupper():
+            raise ValueError("All letters must be uppercase. Got: " + v)
+        return v
+
+
+client = openai.OpenAI()
+try:
+    resp = client.beta.chat.completions.parse(
+        response_format=User,
+        messages=[
+            {
+                "role": "user",
+                "content": "Extract the following user: Jason is 25 years old.",
+            },
+        ],
+        model="gpt-4o-mini",
+    )
+except Exception as e:
+    print(e)
+    """
+    1 validation error for User
+    name
+      Value error, All letters must be uppercase. Got: Jason [type=value_error, input_value='Jason', input_type=str]
+        For further information visit https://errors.pydantic.dev/2.8/v/value_error
+    """
+```
+
+When validation fails, we lose the original completion, making it challenging to implement retry logic or provide specific prompts for correction. This limitation hinders our ability to offer detailed feedback to the model, ultimately impacting our capacity to improve its performance over time.
+
+### Streaming
+
+Streaming with Structured Outputs is supported but a challenging endeavour. There's no built-in partial validation and you need to manually parse the generated response while simultaneously having to now use a context manager to access the generated values.
+
+In short, making it work well in practice requires a good amount of effort with their current `beta.chat.completions.stream` implementation.
+
+```python
+import openai
+from pydantic import BaseModel
+
+
+class User(BaseModel):
+    name: str
+    age: int
+
+
+client = openai.OpenAI()
+with client.beta.chat.completions.stream(
+    response_format=User,
+    messages=[
+        {
+            "role": "user",
+            "content": "Extract the following user: Jason is 25 years old.",
+        },
+    ],
+    model="gpt-4o-mini",
+) as stream:
+    for event in stream:
+        if event.type == "content.delta":
+            print(event.snapshot, flush=True, end="\n")
+            #>
+            #> {"
+            #> {"name
+            #> {"name":"
+            #> {"name":"Jason
+            #> {"name":"Jason","
+            #> {"name":"Jason","age
+            #> {"name":"Jason","age":
+            #> {"name":"Jason","age":25
+            #> {"name":"Jason","age":25}
+```
```

Comment:
bring latency here

Reply:
you also don't mention the fact that we share the same api across many providers so you can still migrate to claude and anthropic  without significant api changes